GVFS/GVFS.Common/Git/GitAuthentication.cs

```diff
@@ -130,5 +207,29 @@ private void UpdateBackoff()
             this.lastAuthAttempt = DateTime.Now;
             this.numberOfAttempts++;
         }
+
+        private bool TryCallGitCredential(ITracer tracer, out string errorMessage)
+        {
+            string gitUsername;
+            string gitPassword;
+            if (!this.git.TryGetCredentials(tracer, this.repoUrl, out gitUsername, out gitPassword, out errorMessage))
+            {
+                this.UpdateBackoff();
+                return false;
+            }
+
+            if (!string.IsNullOrEmpty(gitUsername) && !string.IsNullOrEmpty(gitPassword))
+            {
+                this.cachedAuthString = Convert.ToBase64String(Encoding.ASCII.GetBytes(gitUsername + ":" + gitPassword));
+            }
+            else
+            {
+                // TODO: How can we get back an empty username/password? This may have been an early attempt to "handle" anonymous auth

```

Comment:
> Any idea why we used to allow it?

I checked the history of the old repo and as best I can tell we've allowed this ever since the method was added to support on-demand hydration (back in Feb. 2016).  This makes me think it was an oversight and not intentional.

> From what I can tell, this should be an error.

I agree.

Reply:
Actually, the way we can get back an empty username/password is if we call `git credential`, the GCM fails to generate a PAT and so shows an auth dialog to the user, and the user hits OK without entering anything. In fact, I believe this will work today to clone from a repo that allows anonymous access, it's just annoying because you see that dialog every time you clone, mount, fetch, etc.

So this leaves the question of what we should do about it here. I think now that we're explicitly testing for anonymous access before we can ever interact with the user, getting back an empty string here means that we're going to fail to authenticate anyway, and treating it as an error would make sense. But I'm not 100% confident that that's the end of the story.

cc @jrbriggs any thoughts on this?

Reply:
If I'm understanding the flow correctly, by the time we call this, we have already answered the question of whether the repo allows anonymous access. If we're calling to get credentials for authenticated access and are returned empty strings, it seems like we should treat that as an error.

Reply:
I believe this was to handle NTLM authentication.  Have you tested that use case?

Reply:
@kewillford That's a great point! No I have no environment to test out NTLM auth. Do you have access to one?

Reply:
And @kewillford from an HTTP standpoint, doesn't NTLM just look like anonymous auth? In both cases, we just skip sending the auth header, right? If that's true then we should just fall into anonymous mode when dealing with NTLM, and the user should never see an auth dialog, and so we will still never see an empty username/password in this code path, unless it's an error.

Reply:
Yep see https://github.com/Microsoft/VFSForGit/blob/53def8d0439762cf78905826ad48ceef19907ef2/GVFS/GVFS.Common/Http/HttpRequestor.cs#L93, whether we claim we're "anonymous" or claim we're "authenticated" with an empty username/password, the end result should be the same as far as NTLM is concerned. But it would be awesome if you (or someone!) has an NTLM environment to test it out in.

Reply:
Wil had an onprem server set up before to test against.  I will check with him and see if that is still setup.